client/send-video.cc

```diff
@@ -118,148 +132,167 @@ int main(int argc, char *argv[]) {
+        sws_ctx = sws_getContext(pCodecCtx->width,
+                                 pCodecCtx->height,
+                                 pCodecCtx->pix_fmt,
+                                 display.width(), display.height(),
+                                 AV_PIX_FMT_RGB24,
+                                 SWS_BILINEAR,
+                                 NULL,
+                                 NULL,
+                                 NULL
+                                 );
+        if (sws_ctx == 0) {
+            fprintf(stderr, "Trouble doing scaling to %dx%d :(\n",
+                    display.width(), display.height());
+            return 1;
+        }
 
-    if (off_z > 0) {
-        display.Clear();
-        display.Send();
-    }
+        signal(SIGTERM, InterruptHandler);
+        signal(SIGINT, InterruptHandler);
+
+        // Read frames and send to FlaschenTaschen.
+        const int frame_wait_micros = 1e6 / fps;
+        const tmillis_t startPlay = GetTimeInMillis(); //mark time
+        int repeatedcount = 0;
+        do {
+          frame_count = 0;
+          while (!interrupt_received && av_read_frame(pFormatCtx, &packet) >= 0) {
+              // Is this a packet from the video stream?
+              if (packet.stream_index==videoStream) {
+                  // Decode video frame
+                  avcodec_decode_video2(pCodecCtx, pFrame, &frameFinished, &packet);
+
+                  // Did we get a video frame?
+                  if (frameFinished) {
+                      // Convert the image from its native format to RGB
+                      sws_scale(sws_ctx, (uint8_t const * const *)pFrame->data,
+                                pFrame->linesize, 0, pCodecCtx->height,
+                                pFrameRGB->data, pFrameRGB->linesize);
+
+                      // Save the frame to disk
+                      SendFrame(pFrameRGB, &display);
+                      frame_count++;
+                  }
+                  usleep(frame_wait_micros);
+              }
+
+              // Free the packet that was allocated by av_read_frame
+              av_free_packet(&packet);
+          }
+          if ((GetTimeInMillis() - startPlay) < (repeatSeconds * 1000) && !av_seek_frame(pFormatCtx, -1, 1, AVSEEK_FLAG_FRAME)) {
+              if (repeatedcount)
+                  fprintf(stderr, "finished loop %d (%ld frames) after %0.1fs\n", repeatedcount, frame_count, (GetTimeInMillis() - startPlay) / 1000.);
+              repeatedcount++;
+              continue;
+          } else break;
+        } while (!interrupt_received);
+
+        if (interrupt_received) {
+            // Feedback for Ctrl-C, but most importantly, force a newline
+            // at the output, so that commandline-shell editing is not messed up.
+            fprintf(stderr, "Got interrupt. Exiting\n");
+            return 1;
+        }
 
-    // Free the RGB image
-    av_free(buffer);
-    av_frame_free(&pFrameRGB);
+        if (off_z > 0) {
```

Comment:
Is this something needed here ? This should only be at the very end, when we shut down to leave the z-layer in a clean state.